File: 744. Find Smallest Letter Greater Than Target/nextGreatestLetter.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    char nextGreatestLetter(vector<char>& letters, char target) {
        //non-decreasing => increasing => small -> high
        int left = 0;
        int right = letters.size()-1;
        char ans = letters[0];

        while(left <= right){
            int mid = (left+right) / 2;
            if(letters[mid] > target){
                ans = letters[mid];
                right = mid -1;
                
            }
            else{
                left = mid + 1;
            }
        }
        cout << "ans " << ans<< endl;
        return ans;
    }
};

File: 744. Find Smallest Letter Greater Than Target/nextGreatestLetter.js
/**
 * @param {character[]} letters
 * @param {character} target
 * @return {character}
 */
var nextGreatestLetter = function (letters, target) {
    //non-decreasing => increasing => small -> high
    let left = 0;
    let right = letters.length - 1;
    let ans = letters[0]

    while (left <= right) {
        let mid = Math.floor((left + right) / 2)

        if (letters[mid] > target) {
            ans = letters[mid]
            right = mid - 1
        }
        else {
            left = mid + 1
        }
    }

    return ans;
};